### ISSUE TITLE
Review 00: "Introduction" through "Designing the Core Logic Module"

### ISSUE THREAD

**Issue by mbland** [18F/unit-testing-node#10]
Technical review corresponding to 18F/writing-lab#105.

**Comment by rogeruiz** [comment]
## [Tools](https://pages.18f.gov/unit-testing-node/setup/tools/)

> It's also good practice to apply a linting tool such as JSHint:

We should have the user add `eslint` instead to keep in line with the Compliance
Toolkit. This [package](https://www.npmjs.com/package/gulp-eslint) seems to get
the job done and still align with using `gulp` as the task runner for the project.
The [eslint project is here](https://github.com/eslint/eslint).

cc: @micahsaul on this one since I remember him mentioning that tool during a meeting.

**Comment by rogeruiz** [comment]
## [Designing the Application](https://pages.18f.gov/unit-testing-node/designing-the-application/)

> The high-level flow for the application is:

Since these bullet-points are actually steps in the flow of the application,
they should be an ordered-list instead of an unordered-list.

**Comment by mbland** [comment]
@rogeruiz Thanks for the feedback! I've opened #23 to track this issue, and have already made the ordered list change.

I've also replaced `jshint` with `eslint` in my client and got it to run, but I have a couple of questions. First, to be clear, I've installed [the 18F/compliance-toolkit `.eslintrc`](https://github.com/18F/compliance-toolkit/blob/master/configs/static/.eslintrc) and installed [`eslint-plugin-{scanjs-rules,no-unsafe-innerhtml}`](https://pages.18f.gov/before-you-ship/security/static-analysis/#javascript). I get these warnings from the existing solution code:

``` sh
$ gulp lint-all
[10:22:52] Using gulpfile ~/src/18F/unit-testing-node/gulpfile.js
[10:22:52] Starting 'lint-all'...
[10:22:54]
solutions/02-slack-client/lib/slack-client.js
  59:5  warning  The function setTimeout can be unsafe  scanjs-rules/call_setTimeout

solutions/03-github-client/lib/github-client.js
  56:5  warning  The function setTimeout can be unsafe  scanjs-rules/call_setTimeout

solutions/05-middleware/lib/github-client.js
  56:5  warning  The function setTimeout can be unsafe  scanjs-rules/call_setTimeout

solutions/05-middleware/lib/slack-client.js
  67:5  warning  The function setTimeout can be unsafe  scanjs-rules/call_setTimeout

solutions/06-integration/lib/github-client.js
  56:5  warning  The function setTimeout can be unsafe  scanjs-rules/call_setTimeout

solutions/06-integration/lib/slack-client.js
  67:5  warning  The function setTimeout can be unsafe  scanjs-rules/call_setTimeout

solutions/06-integration/test/integration-test.js
  40:7  warning  The function write can be unsafe  scanjs-rules/call_write

solutions/07-system/lib/github-client.js
  56:5  warning  The function setTimeout can be unsafe  scanjs-rules/call_setTimeout

solutions/07-system/lib/slack-client.js
  67:5  warning  The function setTimeout can be unsafe  scanjs-rules/call_setTimeout

solutions/complete/lib/github-client.js
  56:5  warning  The function setTimeout can be unsafe  scanjs-rules/call_setTimeout

solutions/complete/lib/slack-client.js
  67:5  warning  The function setTimeout can be unsafe  scanjs-rules/call_setTimeout

solutions/complete/test/integration-test.js
  40:7  warning  The function write can be unsafe  scanjs-rules/call_write

✖ 12 problems (0 errors, 12 warnings)

[10:22:54] Finished 'lint-all' after 1.93 s
```

For one, it seems the security scanning rules focus on browser-side JavaScript. Do I need these plugins for server-side, or should I use different ones?

Also, with jshint, it would warn about missing semicolons. Do we have a policy on those? Should I delete all my semicolons? Can/should I update the base `.eshintrc` to add a semicolon rule?

And last one for now: Are we a straight up ES6 shop now? I've been using `Promises`, but have held off on some other handy functions for compatibility's sake.

**Comment by rogeruiz** [comment]
So I believe that the `eslintrc` rules should be agreed upon for the team. In this scenario since it's a guide, I would say that whatever you had in your `jshintrc` would be okay port over to `eslint`. Most rules are off or have specific defaults that may or may not align with the guide automatically.

You can either include an `eslintrc` at the root of a directory for server v. client code with different rules, along with a spec / test directory. I would recommend for this guide that it should align more to the spec / test server-side code.

I'd say we are an ES3-5-6+ shop so most of the code examples I've seen are okay.

For the guide, I think it might just be a good idea to have `eslint` rather than `jshint` since it aligns with the toolkit. Someone could be using the guide to create tests for a node app that might need to go through the process. As far as the rules you've added from the project, they are still in flux so I wouldn't necessarily use them out of the box. 

Feel free to edit the example below as you see fit for the guide. [All the rules are here](http://eslint.org/docs/rules/). Eventually there might be a conversation around adding the toolkit's specific rules and a PR, but having `eslint` in the guides would be great since it wouldn't require migrating from `jshint` to `eslint`.

Here's a bare-bones _opinionated_ `eslintrc` file I use that helps start a conversation around linting with a new team / project. 

```
{
    "rules": {
        "indent": [
            2,
            2
        ],
        "quotes": [
            2,
            "single"
        ],
        "linebreak-style": [
            2,
            "unix"
        ],
        "semi": [
            2,
            "always"
        ],
        "comma-dangle": [
            1,
            "always-multiline"
        ],
        "no-console": [ 0 ]
    },
    "env": {
        "node": true,
        "browser": true
    },
    "extends": "eslint:recommended"
}
```

**Comment by micahsaul** [comment]
Yup, the eslintrc file in Compliance Toolkit is just for the security stuff, and is evolving as we go. It's meant to be used in addition to whatever rules you might be using for style on your project.

**Comment by mbland** [comment]
Just pushed a big commit to #23. It seems the Compliance Toolkit doesn't yet have a server-side `.eslintrc`. Maybe the `.eslintrc` @rogeruiz posted above, or from #23 could be the start of one?

**Comment by mbland** [comment]
I'll also look into adding `brakeman` for the Ruby bits in another PR.
